### DISCUSSION TITLE
Handle multiple Remix-Auth Strategies

### DISCUSSION THREAD

**Discussion by aydrian** [epicweb-dev/epic-stack#65] [Ideas]
Currently it looks like the database is only set up to handle the Form Strategy. I saw in another post someone mention social logins which would need their own strategy. Have you given any thought to how you would handle multiple strategies with Remix-Auth, specifically with the database setup? There's the [Auth.js Adapter approach](https://authjs.dev/reference/adapter/prisma) which might be overly complicated. The `password` property on the `User` model is optional so maybe adding a `strategy` property will be enough.

**Comment by kentcdodds** [comment]
I still need to look into this. I'm definitely planning on supporting multiple in the starter. Probably GitHub. I'm thinking a `strategy` is the way to do it, but I can't say for sure until I do it.

**Reply by sergiodxa** [reply]
All OAuth providers share a basic OAuthProfile interface they conform to and can extend with extra data the provider gives you.

Since the DB has a User schema and it's not using the user from an OAuth provider as the users table it will be super simple to add support for multiple providers, you only need one verify callback for every strategy that does the same thing:

1. Get the email, username, name (if available) and avatar from the OAuthProfile and use it to do a findOrCreate (upsert in Prisma) on the users table.
2. Get that same info but add the user ID on the provider side (also in OAuthProfile), then have a Connections model in the schema (similar to how password works) and use that to do the find or create.

The first approach will use the email (which is already unique) as a way to know who is the user in the app DB, this is great for multiple providers because if a user did the login with GitHub and later used Google as long the email is the same the user is the same.

The second approach allows users to keep multiple connections, this can be more work but it will allow users to have an account in the app for their Google login and another for GitHub, this will need to allow emails to not be unique since the email will be used for more than one connection.

Also the user could from their profile connect more accounts so if you did the creation with password, you can go to the profile and connect Google and GitHub to use them.

If you don't remove the unique in the email then the strategy callback will need to check if an user exists and add the connection row assigned to the user.

---

I think the first approach is the simplest one, but the second will allow users to connect accounts with a different email, which can work for apps like GitHub where the user may have a personal account but needs to connect a work account using their work email too.

It can also be important if the user uses Apple to sign in and decided to hide their email, and later want to also connect another provider but without the hidden email, in that case the user.email could still be the Apple one but also connect a Google/GitHub/etc account

**Reply by kentcdodds** [reply]
I think we'll probably want to go with option 2 because it's the most flexible and isn't actually all that much more complicated. Worth it for the adaptability aspect. Thanks for that rundown @sergiodxa!

**Reply by aydrian** [reply]
@sergiodxa Were you thinking something like this for the provider table? I pulled it from this [OAuth2Profile interface](https://github.com/sergiodxa/remix-auth-oauth2/blob/main/src/index.ts#L16) except for the arrays for emails and photos.

```
model Provider {
  provider    String
  id          String?
  displayName String?
  familyName  String?
  givenName   String?
  middleName  String?
  email       String?
  photo       String?

  user   User   @relation(fields: [userId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  userId String

  @@unique([userId, provider])
}
```

**Reply by sergiodxa** [reply]
More something like this:

```prisma
model Provider {
  provider String
  providerId String

  user   User   @relation(fields: [userId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  userId String

  @@unique([userId, providerIdd])
}
```

The provider is the provider name, the providerId is the ID the user has in that provider, then you relate that to the user. For the unique, you can use either `[userId, provider]` or `[userId, providerId]`, the second way will let a single user connect multiple account of the same provider (e.g. you connect your personal and work Gmail accounts).

The rest of the information can be used to initialize the user profile on your User model, but you don't need to keep the profile per provider.

**Reply by aydrian** [reply]
Ah. Got it. I noticed id is optional in that interface. Do we think every provider will have an id? That would affect which field to use in the unique constraint. I'm leaning toward using provider because every remix-auth provider has a name property that will likely be used for the provider field, right?

**Reply by kentcdodds** [reply]
I'm working on this a bit right now. It's been a while since I implemented OAuth 2 so forgive me, but shouldn't we be storing the refresh token and its expiration date (as well as the access token expiration) somewhere as well?

**Reply by sergiodxa** [reply]
Only if you plan to use the identity provider API.

If you only use (let's say Google or GitHub) to know who is the user (e.g. to store the email, name and an avatar) but you never fetch the Google or GitHub API then you don't need the access token, refresh token or expiration date of them, and for security it would be better to don't store them if you don't need them.

But if you do plan to use their API you will have to store them somewhere, then somehow refresh them. The refresh could work in some ways:

1. Try to use the access token, if expired refresh it, save the new token into the DB and try to use it again.
2. Run a background job that check the expired (or about to expire) tokens and refresh them, so your app can always use the access token supposing they will never be expired.
3. Combine both, so you have the token refreshed in background but your code can refresh and retry if it still fails.

I think for the majority of the apps, they only use OAuth to know who is the user, specially apps with more than one IdP option, but some apps do need to access the API, and in those cases they either have a single IdP option at login/signup or they need to request access later depending on what they plan to use.

**Reply by kentcdodds** [reply]
Ah, that makes plenty of sense. I suppose then that for the user identity case, it doesn't matter whether this option is checked:
<img width="719" alt="image" src="https://github.com/epicweb-dev/epic-stack/assets/1500684/db62c9d0-3afa-4aaa-8c6f-f1275f0767c2">

But it's probably best to leave it checked so the access token can expire automatically.

Thanks for taking the time to explain that!

**Reply by sergiodxa** [reply]
Another thing to consider is how do you plan to use the token.

If you only use the token as the user interacts with the app (e.g. to fetch the repositories on GitHub) you could just keep the token in the session, read it from there and use it directly, once the token expires you could either refresh it or take the user to the login (access token expiration date and session expiration date could be the same in this scenario).

If you plan to use the token in the background (e.g. periodically fetch data from the API) you will need to store the tokens in the DB so you can use the access token and refresh it if needed.

**Comment by kentcdodds** [comment]
If anyone here wants to make an example, that would help a *ton*: https://github.com/epicweb-dev/epic-stack/blob/main/docs/examples.md

**Comment by kentcdodds** [comment]
I've added a new example to the docs. Here it is! https://github.com/kentcdodds/epic-github-auth

**Reply by cr101** [reply]
Thank you. Will this eventually make it into the core or will it stay as an example?

**Reply by kentcdodds** [reply]
Right now I think I'll keep it as an example
